kroxylicious/src/main/java/io/kroxylicious/proxy/internal/admin/RoutingHttpServer.java

```diff
@@ -0,0 +1,126 @@
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+
+import static io.netty.handler.codec.http.HttpHeaderNames.CONNECTION;
+import static io.netty.handler.codec.http.HttpHeaderNames.CONTENT_LENGTH;
+import static io.netty.handler.codec.http.HttpHeaderValues.CLOSE;
+import static io.netty.handler.codec.http.HttpHeaderValues.KEEP_ALIVE;
+import static io.netty.handler.codec.http.HttpHeaderValues.TEXT_PLAIN;
+import static io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR;
+import static io.netty.handler.codec.http.HttpResponseStatus.NOT_FOUND;
+import static io.netty.handler.codec.rtsp.RtspHeaderNames.CONTENT_TYPE;
+
+public class RoutingHttpServer extends SimpleChannelInboundHandler<HttpObject> {
+
+    private final Map<String, Function<HttpRequest, HttpResponse>> routes;
+    private static final Logger LOGGER = LoggerFactory.getLogger(RoutingHttpServer.class);
+
+    public RoutingHttpServer(Map<String, Function<HttpRequest, HttpResponse>> routes) {
+        this.routes = routes;
+    }
+
+    public static RoutingHttpServerBuilder builder() {
+        return new RoutingHttpServerBuilder();
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) {
+        ctx.flush();
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, HttpObject msg) {
+        if (msg instanceof HttpRequest) {
+            HttpRequest req = (HttpRequest) msg;
+            boolean keepAlive = HttpUtil.isKeepAlive(req);
+
+            HttpResponse response = getResponse(req);
+
+            if (keepAlive) {
+                if (!req.protocolVersion().isKeepAliveDefault()) {
+                    response.headers().set(CONNECTION, KEEP_ALIVE);
+                }
+            }
+            else {
+                // Tell the client we're going to close the connection.
+                response.headers().set(CONNECTION, CLOSE);
+            }
+
+            ChannelFuture f = ctx.write(response);
+
+            if (!keepAlive) {
+                f.addListener(ChannelFutureListener.CLOSE);
+            }
+        }
+    }
+
+    private HttpResponse getResponse(HttpRequest req) {
+        if (routes.containsKey(req.uri())) {
+            try {
+                return routes.get(req.uri()).apply(req);
+            }
+            catch (Exception e) {
+                LOGGER.error("exception while invoking endpoint for route {}", req.uri(), e);
+                return responseWithBody(req, "500: internal server error", INTERNAL_SERVER_ERROR);
+            }
+        }
+        else {
+            return responseWithBody(req, "404: Not found", NOT_FOUND);
```

Comment:
Is it necessary to include the status code in the message? It is duplicating the actual status code, and as a convention would allow our future selves to mix up status codes in messages with the actual status code.

Reply:
Introduced a new `responseWithStatus(HttpRequest req, HttpResponseStatus status)` method instead that uses the `reasonPhrase` from the `HttpResponseStatus` instead.